subsys/net/lib/lwm2m/lwm2m_rw_cbor.c

```diff
@@ -0,0 +1,553 @@
+					  char *buf, size_t buflen)
+{
+	int payload_len;
+	bool ret;
+
+	ZCBOR_STATE_E(states, 0, CPKT_BUF_W_PTR(out->out_cpkt), CPKT_BUF_W_SIZE(out->out_cpkt),  1);
+
+	ret = zcbor_tstr_encode(states, &(struct zcbor_string){ .value = buf, .len = buflen });
+
+	if (ret) {
+		payload_len = CPKT_CBOR_W_SZ(states[0].payload, out->out_cpkt);
+		out->out_cpkt->offset += payload_len;
+	} else {
+		payload_len = -ENOMEM;
+	}
+
+	return payload_len;
+}
+
+static int put_opaque(struct lwm2m_output_context *out, struct lwm2m_obj_path *path,
+					  char *buf, size_t buflen)
+{
+	int payload_len;
+	bool ret;
+
+	ZCBOR_STATE_E(states, 0, CPKT_BUF_W_PTR(out->out_cpkt), CPKT_BUF_W_SIZE(out->out_cpkt),  1);
+
+	ret = zcbor_bstr_encode(states, &(struct zcbor_string){ .value = buf, .len = buflen });
+
+	if (ret) {
+		payload_len = CPKT_CBOR_W_SZ(states[0].payload, out->out_cpkt);
+		out->out_cpkt->offset += payload_len;
+	} else {
+		payload_len = -ENOMEM;
+	}
+
+	return payload_len;
+}
+
+static int put_bool(struct lwm2m_output_context *out, struct lwm2m_obj_path *path, bool value)
+{
+	int payload_len;
+	bool ret;
+
+	ZCBOR_STATE_E(states, 0, CPKT_BUF_W_PTR(out->out_cpkt), CPKT_BUF_W_SIZE(out->out_cpkt),  1);
+
+	ret = zcbor_bool_encode(states, &value);
+
+	if (ret) {
+		payload_len = CPKT_CBOR_W_SZ(states[0].payload, out->out_cpkt);
+		out->out_cpkt->offset += payload_len;
+	} else {
+		payload_len = -ENOMEM;
+	}
+
+	return payload_len;
+}
+
+static int put_objlnk(struct lwm2m_output_context *out, struct lwm2m_obj_path *path,
+			 struct lwm2m_objlnk *value)
+{
+	char objlnk[sizeof("65535:65535")] = { 0 };
+
+	snprintk(objlnk, sizeof(objlnk), "%" PRIu16 ":%" PRIu16 "", value->obj_id, value->obj_inst);
+
+	return put_string(out, path, objlnk, strlen(objlnk) + 1);
+}
+
+static int get_s64(struct lwm2m_input_context *in, int64_t *value)
+{
+	ZCBOR_STATE_D(states, 0, ICTX_BUF_R_PTR(in), ICTX_BUF_R_LEFT_SZ(in),  1);
+
+	if (!zcbor_int64_decode(states, value)) {
+		LOG_WRN("unable to decode a 64-bit integer value");
+		return -EBADMSG;
+	}
+
+	int len = ICTX_CBOR_R_SZ(states[0].payload, in);
+
+	in->offset += len;
+
+	return len;
+}
+
+static int get_s32(struct lwm2m_input_context *in, int32_t *value)
+{
+	ZCBOR_STATE_D(states, 0, ICTX_BUF_R_PTR(in), ICTX_BUF_R_LEFT_SZ(in),  1);
+
+	if (!zcbor_int32_decode(states, value)) {
+		LOG_WRN("unable to decode a 32-bit integer value");
+		return -EBADMSG;
+	}
+
+	int len = ICTX_CBOR_R_SZ(states[0].payload, in);
+
+	in->offset += len;
+
+	return len;
+}
+
+static int get_float(struct lwm2m_input_context *in, double *value)
+{
+	ZCBOR_STATE_D(states, 0, ICTX_BUF_R_PTR(in), ICTX_BUF_R_LEFT_SZ(in),  1);
+
+	if (!zcbor_float32_decode(states, (float *)value)) {
+		LOG_DBG("unable to decode a 32-bit floating-point value - trying width 64");
+		if (!zcbor_float64_decode(states, value)) {
+			LOG_ERR("unable to decode a floating-point value");
+			return -EBADMSG;
+		}
```

Comment:
```suggestion
	if (!zcbor_float_decode(states, value)) {
		LOG_ERR("unable to decode a floating-point value");
		return -EBADMSG;
```
I think doing `(float *)value` will give the wrong value, float and double aren't bit-compatible like that.